File: app/src/main/java/com/chaslay/pos/domain/model/GiftCardCode.kt
package com.chaslay.pos.domain.model

/** Parse plain gift-card codes or /gift/{code} URLs from scanner input. */
object GiftCardCode {
    fun parse(raw: String): String {
        val trimmed = raw.trim()
        if (trimmed.isEmpty()) return ""
        if (trimmed.startsWith("http://", ignoreCase = true) ||
            trimmed.startsWith("https://", ignoreCase = true)
        ) {
            val match = Regex("/gift/([^/?#\\s]+)", RegexOption.IGNORE_CASE).find(trimmed)
            if (match != null) return match.groupValues[1].trim()
        }
        val inline = Regex("/gift/([^/?#\\s]+)", RegexOption.IGNORE_CASE).find(trimmed)
        if (inline != null) return inline.groupValues[1].trim()
        return trimmed
    }

    fun qrPayload(code: String): String = parse(code).ifBlank { code.trim() }
}

File: backend/src/lib/gift-card-code.ts
/**
 * Parse scanned / pasted gift-card codes (plain EC-… codes or /gift/{code} URLs).
 */
export function parseGiftCardCode(raw: string): string {
  const trimmed = String(raw || "").trim();
  if (!trimmed) return "";

  try {
    if (/^https?:\/\//i.test(trimmed)) {
      const url = new URL(trimmed);
      const pathMatch = url.pathname.match(/\/gift\/([^/]+)/i);
      if (pathMatch?.[1]) {
        return decodeURIComponent(pathMatch[1]).trim();
      }
    }
  } catch {
    /* not a URL */
  }

  const inlineMatch = trimmed.match(/\/gift\/([^/?#\s]+)/i);
  if (inlineMatch?.[1]) {
    return decodeURIComponent(inlineMatch[1]).trim();
  }

  return trimmed;
}

/** Stable QR / barcode payload — redeem code only (POS scanners). */
export function buildGiftCardRedeemQrPayload(code: string): string {
  return parseGiftCardCode(code) || String(code || "").trim();
}

/** Optional public deep link shown in emails (not required for POS redeem). */
export function buildGiftCardRedeemUrl(code: string): string {
  const parsed = buildGiftCardRedeemQrPayload(code);
  const base = (
    process.env.GIFT_CARD_PUBLIC_BASE_URL ||
    process.env.PUBLIC_APP_URL ||
    "https://app.chaslay.com"
  )
    .replace(/\/$/, "")
    .replace(/^https?:\/\/pay\./i, "https://app.");
  return `${base}/gift/${encodeURIComponent(parsed)}`;
}
